pkg/reconciler/trigger/trigger.go

```diff
@@ -0,0 +1,424 @@
+	ctx = logging.WithLogger(ctx, r.Logger.Desugar().With(zap.String("key", key)))
+
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		r.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the Trigger resource with this namespace/name.
+	original, err := r.triggerLister.Triggers(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("trigger key in work queue no longer exists", zap.Any("key", key))
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy.
+	trigger := original.DeepCopy()
+
+	// Reconcile this copy of the Trigger and then write back any status updates regardless of
+	// whether the reconcile error out.
+	err = r.reconcile(ctx, trigger)
+	if err != nil {
+		logging.FromContext(ctx).Error("Error reconciling Trigger", zap.Error(err))
+		r.Recorder.Eventf(trigger, corev1.EventTypeWarning, triggerReconcileFailed, "Trigger reconciliation failed: %v", err)
+	} else {
+		logging.FromContext(ctx).Debug("Trigger reconciled")
+		r.Recorder.Event(trigger, corev1.EventTypeNormal, triggerReconciled, "Trigger reconciled")
+	}
+
+	if _, updateStatusErr := r.updateStatus(ctx, trigger); updateStatusErr != nil {
+		logging.FromContext(ctx).Error("Failed to update Trigger status", zap.Error(updateStatusErr))
+		r.Recorder.Eventf(trigger, corev1.EventTypeWarning, triggerUpdateStatusFailed, "Failed to update Trigger's status: %v", err)
+		return updateStatusErr
+	}
+
+	// Requeue if the resource is not ready
+	return err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, t *v1alpha1.Trigger) error {
+	t.Status.InitializeConditions()
+
+	// 1. Verify the Broker exists.
+	// 2. Get the Broker's:
+	//   - Trigger Channel
+	//   - Ingress Channel
+	//   - Filter Service
+	// 3. Find the Subscriber's URI.
+	// 4. Creates a Subscription from the Broker's Trigger Channel to this Trigger via the Broker's
+	//    Filter Service with a specific path, and reply set to the Broker's Ingress Channel.
+
+	if t.DeletionTimestamp != nil {
+		// Everything is cleaned up by the garbage collector.
+		return nil
+	}
+
+	b, err := r.brokerLister.Brokers(t.Namespace).Get(t.Spec.Broker)
+	if err != nil {
+		logging.FromContext(ctx).Error("Unable to get the Broker", zap.Error(err))
+		if apierrs.IsNotFound(err) {
+			t.Status.MarkBrokerFailed("DoesNotExist", "Broker does not exist")
+		} else {
+			t.Status.MarkBrokerFailed("BrokerGetFailed", "Failed to get broker")
+		}
+		return err
+	}
+	t.Status.PropagateBrokerStatus(&b.Status)
+	logging.FromContext(ctx).Error("Setting status to", zap.Any("status:", t.Status))
+
+	// Tell tracker to reconcile this Trigger whenever the Broker changes.
+	gvk := v1alpha1.SchemeGroupVersion.WithKind("Broker")
+	if err = r.tracker.Track(objectRef(b, gvk), t); err != nil {
```

Comment:
Does track based on name or UID? Specifically if this Trigger is tracking a Broker. That Broker is deleted. A new Broker with the same name is created. Will this Trigger automatically track that new Broker? Or will it need to wait for the Trigger to reconcile for some other reason, at which point it tracks the new one?

Reply:
yes, it tracks by the name. I'll try to come up with tests for this in a follow up.